### ISSUE TITLE
afterSlide never called

### ISSUE THREAD

**Issue by ms-tng** [FormidableLabs/nuka-carousel#235]
The afterSlide handler is never called.

The reason is in carousel.js, line 503:
`if (index !== this.state.currentSlide) {`
Here, `index`and `this.state.currentSlide` are always equal since the comparison occurs in a setState callback that gets called immediately after `this.state.currentSlide` is set to `index`.

**Comment by LiHwsqh** [comment]
@kenwheeler let's just fix it

`   if (index !== this.state.currentSlide) {
      this.props.afterSlide(index);
    }

    this.setState({
      currentSlide: index
    }, function () {
      self.animateSlide();
      self.resetAutoplay();
      self.setExternalData();
    });
`

**Comment by ms-tng** [comment]
@LiHwsqh That would mean calling afterSlide **before** the slide occurs.
Also, before fixing it there should be a red test.

**Comment by william-abboud** [comment]
You can use the `beforeSlide` function to implement your logic. Its being called right before line 499 and has no weird guard checks:
```
this.props.beforeSlide(this.state.currentSlide, index);

    this.setState({
      currentSlide: index
    }, function () {
      self.animateSlide();
      if (index !== this.state.currentSlide) { // <- This is always false
        this.props.afterSlide(index);
      }
      self.resetAutoplay();
      self.setExternalData();
    });

```

**Comment by ms-tng** [comment]
@william-abboud Thanks. It's actually not causing me trouble right now. I do need both beforeSlide and afterSlide, but I can stick with the previous version where it worked fine.

It should still be fixed, though, and that's why I opened the issue.

**Comment by william-abboud** [comment]
Agreed @ms-tng ! Should be fixed.

**Comment by ms-tng** [comment]
@kenwheeler Commit fb3f5fe looks like it is supposed to fix this issue. However, now both beforeSlide and afterSlide get called before the slide happens.

**Comment by LiHwsqh** [comment]
@ms-tng @kenwheeler   Maybe we can save the index before it changes
```javascript

    const slideDoesChange = index === this.state.currentSlide;

    this.setState({
      currentSlide: index
    }, function () {
      self.animateSlide();
      self.resetAutoplay();
      self.setExternalData();
      if (slideDoesChange) {
        this.props.afterSlide(index);
      }
    });
```

**Comment by ms-tng** [comment]
@LiHwsqh @kenwheeler Looks good. There should still be a red test first, however.

**Comment by sarmeyer** [comment]
Fixed in PR #396 (re: issue #376)
